File: order/src/test/java/com/yas/order/service/OrderServiceTest.java
package com.yas.order.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.yas.commonlibrary.exception.NotFoundException;
import com.yas.order.mapper.OrderMapper;
import com.yas.order.model.Order;
import com.yas.order.model.OrderAddress;
import com.yas.order.model.OrderItem;
import com.yas.order.model.csv.OrderItemCsv;
import com.yas.order.model.enumeration.DeliveryMethod;
import com.yas.order.model.enumeration.DeliveryStatus;
import com.yas.order.model.enumeration.OrderStatus;
import com.yas.order.model.enumeration.PaymentMethod;
import com.yas.order.model.enumeration.PaymentStatus;
import com.yas.order.model.request.OrderRequest;
import com.yas.order.repository.OrderItemRepository;
import com.yas.order.repository.OrderRepository;
import com.yas.order.viewmodel.order.OrderBriefVm;
import com.yas.order.viewmodel.order.OrderExistsByProductAndUserGetVm;
import com.yas.order.viewmodel.order.OrderItemPostVm;
import com.yas.order.viewmodel.order.OrderListVm;
import com.yas.order.viewmodel.order.OrderPostVm;
import com.yas.order.viewmodel.order.OrderVm;
import com.yas.order.viewmodel.order.PaymentOrderStatusVm;
import com.yas.order.viewmodel.orderaddress.OrderAddressPostVm;
import com.yas.order.viewmodel.orderaddress.OrderAddressVm;
import com.yas.order.viewmodel.promotion.PromotionUsageVm;
import java.io.IOException;
import java.math.BigDecimal;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.data.util.Pair;

@ExtendWith(MockitoExtension.class)
class OrderServiceTest {

    @Mock
    private OrderRepository orderRepository;

    @Mock
    private OrderItemRepository orderItemRepository;

    @Mock
    private ProductService productService;

    @Mock
    private CartService cartService;

    @Mock
    private OrderMapper orderMapper;

    @Mock
    private PromotionService promotionService;

    @InjectMocks
    private OrderService orderService;

    private Order order;

    @BeforeEach
    void setUp() {
        order = buildOrder(1L, "checkout-1", OrderStatus.PENDING, PaymentStatus.PENDING);
    }

    @Test
    void createOrder_whenValidPayload_shouldPersistAndTriggerDependentActions() {
        OrderPostVm postVm = buildOrderPostVm();

        when(orderRepository.save(any(Order.class))).thenAnswer(invocation -> {
            Order saved = invocation.getArgument(0);
            saved.setId(1L);
            return saved;
        });
        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
        when(orderRepository.save(order)).thenReturn(order);

        OrderVm result = orderService.createOrder(postVm);

        assertThat(result.id()).isEqualTo(1L);
        assertThat(result.orderItemVms()).hasSize(2);
        verify(productService).subtractProductStockQuantity(any(OrderVm.class));
        verify(cartService).deleteCartItems(any(OrderVm.class));
        verify(promotionService).updateUsagePromotion(any(List.class));
        verify(orderRepository, times(2)).save(any(Order.class));
    }

    @Test
    void getOrderWithItemsById_whenFound_shouldReturnVmWithItems() {
        List<OrderItem> orderItems = List.of(
                buildOrderItem(100L, 1L, "item-a", 2),
                buildOrderItem(101L, 1L, "item-b", 1)
        );
        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
        when(orderItemRepository.findAllByOrderId(1L)).thenReturn(orderItems);

        OrderVm result = orderService.getOrderWithItemsById(1L);

        assertThat(result.id()).isEqualTo(1L);
        assertThat(result.orderItemVms()).hasSize(2);
    }

    @Test
    void getOrderWithItemsById_whenNotFound_shouldThrowNotFoundException() {
        when(orderRepository.findById(anyLong())).thenReturn(Optional.empty());

        assertThatThrownBy(() -> orderService.getOrderWithItemsById(99L))
                .isInstanceOf(NotFoundException.class);
    }

    @Test
    void getAllOrder_whenEmptyPage_shouldReturnEmptyResult() {
        when(orderRepository.findAll(any(Specification.class), any(Pageable.class)))
                .thenReturn(Page.empty());

        OrderListVm result = orderService.getAllOrder(
                Pair.of(ZonedDateTime.now().minusDays(1), ZonedDateTime.now()),
                "phone",
                List.of(),
                Pair.of("VN", "0123"),
                "customer@example.com",
                Pair.of(0, 10)
        );

        assertThat(result.orderList()).isNull();
        assertThat(result.totalElements()).isZero();
        assertThat(result.totalPages()).isZero();
    }

    @Test
    void getAllOrder_whenPageHasData_shouldMapToOrderBriefVm() {
        Page<Order> orderPage = new PageImpl<>(List.of(order));
        when(orderRepository.findAll(any(Specification.class), any(Pageable.class)))
                .thenReturn(orderPage);

        OrderListVm result = orderService.getAllOrder(
                Pair.of(ZonedDateTime.now().minusDays(1), ZonedDateTime.now()),
                "phone",
                List.of(OrderStatus.PENDING),
                Pair.of("VN", "0123"),
                "customer@example.com",
                Pair.of(0, 10)
        );

        assertThat(result.orderList()).hasSize(1);
        assertThat(result.totalElements()).isEqualTo(1);
        assertThat(result.totalPages()).isEqualTo(1);
    }

    @Test
    void getLatestOrders_whenCountLessThanOrEqualZero_shouldReturnEmptyList() {
        List<OrderBriefVm> result = orderService.getLatestOrders(0);

        assertThat(result).isEmpty();
    }

    @Test
    void getLatestOrders_whenRepositoryReturnsEmpty_shouldReturnEmptyList() {
        when(orderRepository.getLatestOrders(any(Pageable.class))).thenReturn(List.of());

        List<OrderBriefVm> result = orderService.getLatestOrders(5);

        assertThat(result).isEmpty();
    }

    @Test
    void getLatestOrders_whenRepositoryReturnsOrders_shouldReturnMappedItems() {
        when(orderRepository.getLatestOrders(any(Pageable.class))).thenReturn(List.of(order));

        List<OrderBriefVm> result = orderService.getLatestOrders(5);

        assertThat(result).hasSize(1);
        assertThat(result.get(0).id()).isEqualTo(1L);
    }

    @Test
    void findOrderByCheckoutId_whenFound_shouldReturnOrder() {
        when(orderRepository.findByCheckoutId("checkout-1")).thenReturn(Optional.of(order));

        Order result = orderService.findOrderByCheckoutId("checkout-1");

        assertThat(result.getId()).isEqualTo(1L);
    }

    @Test
    void findOrderByCheckoutId_whenNotFound_shouldThrowNotFoundException() {
        when(orderRepository.findByCheckoutId("missing")).thenReturn(Optional.empty());

        assertThatThrownBy(() -> orderService.findOrderByCheckoutId("missing"))
                .isInstanceOf(NotFoundException.class);
    }

    @Test
    void findOrderVmByCheckoutId_whenFound_shouldReturnOrderVm() {
        List<OrderItem> orderItems = List.of(buildOrderItem(100L, 1L, "item-a", 2));
        when(orderRepository.findByCheckoutId("checkout-1")).thenReturn(Optional.of(order));
        when(orderItemRepository.findAllByOrderId(1L)).thenReturn(orderItems);

        var result = orderService.findOrderVmByCheckoutId("checkout-1");

        assertThat(result.id()).isEqualTo(1L);
        assertThat(result.orderItems()).hasSize(1);
    }

    @Test
    void updateOrderPaymentStatus_whenCompleted_shouldMarkOrderAsPaid() {
        PaymentOrderStatusVm request = PaymentOrderStatusVm.builder()
                .orderId(1L)
                .paymentId(100L)
                .paymentStatus(PaymentStatus.COMPLETED.name())
                .build();

        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
        when(orderRepository.save(order)).thenReturn(order);

        PaymentOrderStatusVm result = orderService.updateOrderPaymentStatus(request);

        assertThat(order.getOrderStatus()).isEqualTo(OrderStatus.PAID);
        assertThat(order.getPaymentStatus()).isEqualTo(PaymentStatus.COMPLETED);
        assertThat(result.orderId()).isEqualTo(1L);
        assertThat(result.paymentId()).isEqualTo(100L);
    }

    @Test
    void updateOrderPaymentStatus_whenNonCompleted_shouldNotChangeOrderStatusToPaid() {
        PaymentOrderStatusVm request = PaymentOrderStatusVm.builder()
                .orderId(1L)
                .paymentId(101L)
                .paymentStatus(PaymentStatus.PENDING.name())
                .build();

        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
        when(orderRepository.save(order)).thenReturn(order);

        orderService.updateOrderPaymentStatus(request);

        assertThat(order.getOrderStatus()).isEqualTo(OrderStatus.PENDING);
        assertThat(order.getPaymentStatus()).isEqualTo(PaymentStatus.PENDING);
    }

    @Test
    void rejectOrder_whenFound_shouldSetRejectStatusAndReason() {
        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));

        orderService.rejectOrder(1L, "out-of-stock");

        assertThat(order.getOrderStatus()).isEqualTo(OrderStatus.REJECT);
        assertThat(order.getRejectReason()).isEqualTo("out-of-stock");
        verify(orderRepository).save(order);
    }

    @Test
    void acceptOrder_whenFound_shouldSetAcceptedStatus() {
        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));

        orderService.acceptOrder(1L);

        assertThat(order.getOrderStatus()).isEqualTo(OrderStatus.ACCEPTED);
        verify(orderRepository).save(order);
    }

    @Test
    void exportCsv_whenOrderListIsNull_shouldReturnCsvBytes() throws IOException {
        OrderService serviceSpy = org.mockito.Mockito.spy(orderService);
        OrderRequest request = buildOrderRequest();
        doReturn(new OrderListVm(null, 0, 0)).when(serviceSpy)
            .getAllOrder(any(), any(), any(), any(), any(), any());

        byte[] result = serviceSpy.exportCsv(request);

        assertThat(result).isNotNull();
        assertThat(result.length).isGreaterThan(0);
    }

    @Test
    void exportCsv_whenOrderListHasData_shouldUseOrderMapper() throws IOException {
        OrderService serviceSpy = org.mockito.Mockito.spy(orderService);
        OrderRequest request = buildOrderRequest();
        OrderBriefVm briefVm = OrderBriefVm.builder()
                .id(1L)
                .email("customer@example.com")
                .billingAddressVm(OrderAddressVm.builder().phone("0123").build())
                .orderStatus(OrderStatus.PENDING)
                .paymentStatus(PaymentStatus.PENDING)
                .deliveryStatus(DeliveryStatus.PREPARING)
                .totalPrice(BigDecimal.TEN)
                .createdOn(ZonedDateTime.now())
                .build();
        doReturn(new OrderListVm(List.of(briefVm), 1, 1)).when(serviceSpy)
            .getAllOrder(any(), any(), any(), any(), any(), any());
        when(orderMapper.toCsv(briefVm)).thenReturn(OrderItemCsv.builder().email("customer@example.com").build());

        byte[] result = serviceSpy.exportCsv(request);

        assertThat(result).isNotNull();
        assertThat(result.length).isGreaterThan(0);
        verify(orderMapper).toCsv(briefVm);
    }

    @Test
    void createOrder_shouldCreatePromotionUsageEntriesForOrderItems() {
        OrderPostVm postVm = buildOrderPostVm();

        when(orderRepository.save(any(Order.class))).thenAnswer(invocation -> {
            Order saved = invocation.getArgument(0);
            saved.setId(1L);
            return saved;
        });
        when(orderRepository.findById(1L)).thenReturn(Optional.of(order));
        when(orderRepository.save(order)).thenReturn(order);

        orderService.createOrder(postVm);

        ArgumentCaptor<List<PromotionUsageVm>> captor = ArgumentCaptor.forClass(List.class);
        verify(promotionService).updateUsagePromotion(captor.capture());
        assertThat(captor.getValue()).hasSize(2);
    }

    private OrderPostVm buildOrderPostVm() {
        OrderAddressPostVm address = OrderAddressPostVm.builder()
                .contactName("John")
                .phone("0123456789")
                .addressLine1("123 Main St")
                .addressLine2("Apt 10")
                .city("HCM")
                .zipCode("70000")
                .districtId(1L)
                .districtName("D1")
                .stateOrProvinceId(79L)
                .stateOrProvinceName("HCM")
                .countryId(84L)
                .countryName("Vietnam")
                .build();

        List<OrderItemPostVm> items = List.of(
                OrderItemPostVm.builder()
                        .productId(100L)
                        .productName("Product A")
                        .quantity(2)
                        .productPrice(BigDecimal.valueOf(99_000))
                        .note("note a")
                        .build(),
                OrderItemPostVm.builder()
                        .productId(101L)
                        .productName("Product B")
                        .quantity(1)
                        .productPrice(BigDecimal.valueOf(55_000))
                        .note("note b")
                        .build()
        );

        return OrderPostVm.builder()
                .checkoutId("checkout-1")
                .email("customer@example.com")
                .shippingAddressPostVm(address)
                .billingAddressPostVm(address)
                .note("deliver in office hours")
                .tax(1.2f)
                .discount(0.3f)
                .numberItem(3)
                .totalPrice(BigDecimal.valueOf(253_000))
                .deliveryFee(BigDecimal.valueOf(15_000))
                .couponCode("SPRING")
                .deliveryMethod(DeliveryMethod.GRAB_EXPRESS)
                .paymentMethod(PaymentMethod.COD)
                .paymentStatus(PaymentStatus.PENDING)
                .orderItemPostVms(items)
                .build();
    }

    private Order buildOrder(Long id, String checkoutId, OrderStatus orderStatus, PaymentStatus paymentStatus) {
        OrderAddress shippingAddress = OrderAddress.builder()
                .id(11L)
                .phone("0987654321")
                .contactName("Ship")
                .addressLine1("Ship Street")
                .city("HCM")
                .zipCode("70000")
                .districtId(1L)
                .districtName("D1")
                .stateOrProvinceId(79L)
                .stateOrProvinceName("HCM")
                .countryId(84L)
                .countryName("Vietnam")
                .build();

        OrderAddress billingAddress = OrderAddress.builder()
                .id(12L)
                .phone("0123123123")
                .contactName("Bill")
                .addressLine1("Bill Street")
                .city("HCM")
                .zipCode("70000")
                .districtId(1L)
                .districtName("D1")
                .stateOrProvinceId(79L)
                .stateOrProvinceName("HCM")
                .countryId(84L)
                .countryName("Vietnam")
                .build();

        return Order.builder()
                .id(id)
                .email("customer@example.com")
                .checkoutId(checkoutId)
                .note("note")
                .tax(1.1f)
                .discount(0.4f)
                .numberItem(2)
                .totalPrice(BigDecimal.valueOf(100_000))
                .deliveryFee(BigDecimal.valueOf(10_000))
                .couponCode("CODE")
                .orderStatus(orderStatus)
                .deliveryMethod(DeliveryMethod.GRAB_EXPRESS)
                .deliveryStatus(DeliveryStatus.PREPARING)
                .paymentStatus(paymentStatus)
                .shippingAddressId(shippingAddress)
                .billingAddressId(billingAddress)
                .build();
    }

    private OrderItem buildOrderItem(Long productId, Long orderId, String name, int quantity) {
        return OrderItem.builder()
                .productId(productId)
                .orderId(orderId)
                .productName(name)
                .quantity(quantity)
                .productPrice(BigDecimal.valueOf(10_000))
                .note("note")
                .build();
    }

    private OrderRequest buildOrderRequest() {
        return OrderRequest.builder()
                .createdFrom(ZonedDateTime.now().minusDays(1))
                .createdTo(ZonedDateTime.now())
                .warehouse("main")
                .productName("Product")
                .orderStatus(List.of(OrderStatus.PENDING))
                .billingPhoneNumber("0123")
                .email("customer@example.com")
                .billingCountry("VN")
                .pageNo(0)
                .pageSize(10)
                .build();
    }
}

File: backoffice/test/modules/order/services/OrderService.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import {
  getOrders,
  getLatestOrders,
  getOrderById,
} from '../../../../modules/order/services/OrderService';
import { Order } from '../../../../modules/order/models/Order';
import { OrderItem } from '../../../../modules/order/models/OrderItem';
import { OrderAddress } from '../../../../modules/order/models/OrderAddress';

// Mock ApiClientService
vi.mock('../../../../common/services/ApiClientService', () => ({
  default: {
    get: vi.fn(),
    post: vi.fn(),
    put: vi.fn(),
    delete: vi.fn(),
  },
}));

import apiClientService from '../../../../common/services/ApiClientService';

describe('OrderService', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  // Mock address data
  const mockAddress: OrderAddress = {
    contactName: 'John Doe',
    phone: '123-456-7890',
    addressLine1: '123 Main St',
    addressLine2: 'Apt 4B',
    zipCode: '10001',
    districtName: 'Manhattan',
    districtId: 1,
    city: 'New York',
    stateOrProvinceName: 'New York',
    stateOrProvinceId: 1,
    countryName: 'United States',
    countryId: 1,
  };

  // Mock order items
  const mockOrderItems: OrderItem[] = [
    {
      id: 1,
      productId: 100,
      productName: 'Laptop',
      quantity: 1,
      productPrice: 999.99,
    },
    {
      id: 2,
      productId: 101,
      productName: 'Mouse',
      quantity: 2,
      productPrice: 29.99,
    },
  ];

  // Mock order data
  const mockOrder: Order = {
    id: 1,
    email: 'john@example.com',
    note: 'Please leave at front door',
    tax: 50.00,
    discount: 10.00,
    numberItem: 3,
    totalPrice: 1059.97,
    deliveryFee: 15.00,
    couponCode: 'SAVE10',
    deliveryMethod: 'Standard Shipping',
    deliveryStatus: 'Shipped',
    paymentMethod: 'Credit Card',
    paymentStatus: 'Paid',
    createdOn: new Date('2024-01-15T10:30:00Z'),
    orderStatus: 'Processing',
    orderItemVms: mockOrderItems,
    shippingAddressVm: mockAddress,
    billingAddressVm: mockAddress,
    checkoutId: 'checkout_123',
  };

  describe('getOrders', () => {
    const mockParams = 'pageNo=1&pageSize=10&sort=id,desc';

    it('should call API with correct URL', async () => {
      const mockResponse = {
        orderList: [mockOrder],
        totalPages: 5,
        totalElements: 50,
      };
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockResponse) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrders(mockParams);

      expect(apiClientService.get).toHaveBeenCalledTimes(1);
      expect(apiClientService.get).toHaveBeenCalledWith(
        '/api/order/backoffice/orders?pageNo=1&pageSize=10&sort=id,desc'
      );
    });

    it('should return orders data on success', async () => {
      const mockResponse = {
        orderList: [mockOrder],
        totalPages: 5,
        totalElements: 50,
      };
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockResponse) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      const result = await getOrders(mockParams);

      expect(result).toEqual(mockResponse);
      expect(result.orderList).toHaveLength(1);
      expect(result.totalPages).toBe(5);
      expect(result.totalElements).toBe(50);
    });

    it('should handle empty params string', async () => {
      const mockResponse = { orderList: [], totalPages: 0, totalElements: 0 };
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockResponse) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrders('');

      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders?');
    });

    it('should handle empty order list', async () => {
      const mockResponse = { orderList: [], totalPages: 0, totalElements: 0 };
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockResponse) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      const result = await getOrders(mockParams);

      expect(result.orderList).toEqual([]);
      expect(result.totalPages).toBe(0);
    });

    it('should handle API error', async () => {
      const error = new Error('Network error');
      (apiClientService.get as any).mockRejectedValue(error);

      await expect(getOrders(mockParams)).rejects.toThrow('Network error');
    });

    it('should handle complex params string', async () => {
      const complexParams = 'pageNo=2&pageSize=20&sort=createdOn,asc&status=COMPLETED';
      const mockApiResponse = { json: vi.fn().mockResolvedValue({ orderList: [], totalPages: 0, totalElements: 0 }) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrders(complexParams);

      expect(apiClientService.get).toHaveBeenCalledWith(
        '/api/order/backoffice/orders?pageNo=2&pageSize=20&sort=createdOn,asc&status=COMPLETED'
      );
    });
  });

  describe('getLatestOrders', () => {
    const count = 5;

    it('should call API with correct URL', async () => {
      const mockResponse = [mockOrder];
      const mockApiResponse = { 
        status: 200,
        json: vi.fn().mockResolvedValue(mockResponse),
        statusText: 'OK',
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getLatestOrders(count);

      expect(apiClientService.get).toHaveBeenCalledTimes(1);
      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders/latest/5');
    });

    it('should return orders on success (status 200)', async () => {
      const mockResponse = [mockOrder];
      const mockApiResponse = { 
        status: 200,
        json: vi.fn().mockResolvedValue(mockResponse),
        statusText: 'OK',
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      const result = await getLatestOrders(count);

      expect(result).toEqual(mockResponse);
      expect(result).toHaveLength(1);
    });

    it('should return orders on success (status 201)', async () => {
      const mockResponse = [mockOrder];
      const mockApiResponse = { 
        status: 201,
        json: vi.fn().mockResolvedValue(mockResponse),
        statusText: 'Created',
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      const result = await getLatestOrders(count);

      expect(result).toEqual(mockResponse);
    });

    it('should return orders on success (status 204)', async () => {
      const mockResponse = [mockOrder];
      const mockApiResponse = { 
        status: 204,
        json: vi.fn().mockResolvedValue(mockResponse),
        statusText: 'No Content',
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      const result = await getLatestOrders(count);

      expect(result).toEqual(mockResponse);
    });

    it('should reject with error when status is 400', async () => {
      const mockApiResponse = { 
        status: 400,
        statusText: 'Bad Request',
        json: vi.fn(),
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await expect(getLatestOrders(count)).rejects.toThrow('Bad Request');
    });

    it('should reject with error when status is 404', async () => {
      const mockApiResponse = { 
        status: 404,
        statusText: 'Not Found',
        json: vi.fn(),
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await expect(getLatestOrders(count)).rejects.toThrow('Not Found');
    });

    it('should reject with error when status is 500', async () => {
      const mockApiResponse = { 
        status: 500,
        statusText: 'Internal Server Error',
        json: vi.fn(),
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await expect(getLatestOrders(count)).rejects.toThrow('Internal Server Error');
    });

    it('should handle count = 0', async () => {
      const mockApiResponse = { 
        status: 200,
        json: vi.fn().mockResolvedValue([]),
        statusText: 'OK',
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getLatestOrders(0);

      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders/latest/0');
    });

    it('should handle count = 10', async () => {
      const mockApiResponse = { 
        status: 200,
        json: vi.fn().mockResolvedValue([]),
        statusText: 'OK',
      };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getLatestOrders(10);

      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders/latest/10');
    });

    it('should handle API network error', async () => {
      const error = new Error('Network error');
      (apiClientService.get as any).mockRejectedValue(error);

      await expect(getLatestOrders(count)).rejects.toThrow('Network error');
    });
  });

  describe('getOrderById', () => {
    const orderId = 1;

    it('should call API with correct URL', async () => {
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockOrder) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrderById(orderId);

      expect(apiClientService.get).toHaveBeenCalledTimes(1);
      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders/1');
    });

    it('should return order data on success', async () => {
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockOrder) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      const result = await getOrderById(orderId);

      expect(result).toEqual(mockOrder);
      expect(result.id).toBe(1);
      expect(result.email).toBe('john@example.com');
    });

    it('should handle orderId = 0', async () => {
      const mockApiResponse = { json: vi.fn().mockResolvedValue(null) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrderById(0);

      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders/0');
    });

    it('should handle API error when order not found', async () => {
      const error = new Error('Order not found');
      (apiClientService.get as any).mockRejectedValue(error);

      await expect(getOrderById(999)).rejects.toThrow('Order not found');
    });

    it('should handle large order ID', async () => {
      const mockApiResponse = { json: vi.fn().mockResolvedValue({ ...mockOrder, id: 999999 }) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrderById(999999);

      expect(apiClientService.get).toHaveBeenCalledWith('/api/order/backoffice/orders/999999');
    });
  });

  describe('URL construction', () => {
    it('should use correct base URL for getOrders', async () => {
      const mockApiResponse = { json: vi.fn().mockResolvedValue({ orderList: [], totalPages: 0, totalElements: 0 }) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrders('pageNo=1');

      expect(apiClientService.get).toHaveBeenCalledWith(
        expect.stringContaining('/api/order/backoffice/orders')
      );
    });

    it('should use correct base URL for getLatestOrders', async () => {
      const mockApiResponse = { status: 200, json: vi.fn().mockResolvedValue([]), statusText: 'OK' };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getLatestOrders(5);

      expect(apiClientService.get).toHaveBeenCalledWith(
        expect.stringContaining('/api/order/backoffice/orders/latest/5')
      );
    });

    it('should use correct base URL for getOrderById', async () => {
      const mockApiResponse = { json: vi.fn().mockResolvedValue(mockOrder) };
      (apiClientService.get as any).mockResolvedValue(mockApiResponse);

      await getOrderById(1);

      expect(apiClientService.get).toHaveBeenCalledWith(
        expect.stringContaining('/api/order/backoffice/orders/1')
      );
    });
  });
});